### PR TITLE
squashfuse: init at 0.1.101

### PR DESCRIPTION
**Pull request by bignaux**
###### Motivation for this change

new package, latest release.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by grahamc** [comment]
@GrahamcOfBorg build squashfuse

File: pkgs/tools/filesystems/squashfuse/default.nix [hunk 1]
```diff
@@ -0,0 +1,65 @@
+{ stdenv, fetchurl, automake, autoconf, libtool, fuse, pkgconfig, pcre,
+
+# Optional Dependencies
+lz4 ? null, xz ? null, zlib ? null, lzo ? null, zstd ? null}:
```
Comment (Profpatsch) [inline pkgs/tools/filesystems/squashfuse/default.nix]: I’d change that to a list of supported compression algorithms

```
lz4, xz, zlib, lzo, zstd,
compressionAlgorithmSupport ? [ "lz4" "lzma" "zlib" "lzo" "zstd" ]`
```

File: pkgs/tools/filesystems/squashfuse/default.nix [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+with stdenv.lib;
+let
```
Comment (Profpatsch) [inline pkgs/tools/filesystems/squashfuse/default.nix]:
```nix
let
  compressionAlgorithms = {
    inherit lz4 zlib lzo zstd;
    lzma = xz;
  }
```

File: pkgs/tools/filesystems/squashfuse/default.nix [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  mkFlag = trueStr: falseStr: cond: name: val: "--"
+    + (if cond then trueStr else falseStr)
+    + name
+    + optionalString (val != null && cond != false) "=${val}";
+  mkEnable = mkFlag "enable-" "disable-";
+  mkWith = mkFlag "with-" "--without-";
+  mkOther = mkFlag "" "" true;
```
Comment (Profpatsch) [inline pkgs/tools/filesystems/squashfuse/default.nix]: You don’t use these.

File: pkgs/tools/filesystems/squashfuse/default.nix [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	    sha256 = "08d1j1a73dhhypbk0q20qkrz564zpmvkpk3k3s8xw8gd9nvy2xa2";
+	  };
+
+  nativeBuildInputs = [ automake autoconf libtool pkgconfig];
+  buildInputs = [ optLz4 optLzma optZlib optLzo optZstd fuse ];
+
+	# We can do it far better i guess, ignoring -with option
+	# but it should be safer like that.
+	# TODO: Improve writing nix expression mkWithLib.
+  configureFlags = [
```
Comment (Profpatsch) [inline pkgs/tools/filesystems/squashfuse/default.nix]:
```nix
let compressionAlgorithmFlags = with stdenv.lib;
      let mkFlag = libname:
        assert compressionAlgorithms ? libname; # your compression is not supported, see compressionAlgorithms
        "${libname}=${getLib compressionAlgorithms.libname}/lib";
      in map mkFlag compressionAlgorithmSupport;
```

(written from muscle memory, something along that line)

**Review by Profpatsch — CHANGES_REQUESTED** [review]
Instead of playing with `null` everywhere, just import all possible libraries (which won’t get used if they are not needed, nix is lazy) and a list of which compression algorithms should be supported.

Then assign each supported algorithm string to its package and when actually setting `configureFlags`, map everything over the list of strings and use the `compressionAlgorithm` map to construct the flag string (erroring out, if a given string is not supported).

Please comment if the code is not clear. I have written it from the top of my head and not actually evaluated.

File: pkgs/tools/filesystems/squashfuse/default.nix (hunk 1 above)
Comment (Mic92) [reply pkgs/tools/filesystems/squashfuse/default.nix]: looks more complicated actually.

**Comment by Mic92** [comment]
I leave this feature list for compression allgorithms for future pull requests, if somebody is interested in.

File: pkgs/tools/filesystems/squashfuse/default.nix (hunk 1 above)
Comment (Profpatsch) [reply pkgs/tools/filesystems/squashfuse/default.nix]: > Also now nobody can now use the compressionAlgorithmSupport attribute in all-packages without colliding with this package.

I’m not sure I understand?

**Comment by edolstra** [comment]
Please don't reintroduce that `mkFlag` / `shouldUsePkg` stuff. I once spent a whole day getting rid of that stuff (see https://github.com/NixOS/nixpkgs/commits/5d8fe045b6376851c33a1c772cfe46882411cd21) and don't want to do it again.

See https://github.com/NixOS/nixpkgs/commit/25a148fa196b944b3f134527da87e43d88c066f9#commitcomment-11464463 for objections to the `mkFlag` / `shouldUsePkg` style.

**Comment by bignaux** [comment]
@edolstra : I first use it because i found it in another package and it is the most near feature i was looking for, since i'm coming from gentoo. On that, i can make my package depend on specific use flag, and use flag globally. I use squashfuse on some kind of application bundle, and if i add an useless compression in my squashfuse libraries, it takes more space and dependencies to embed in the bundle. So i don't want "import all possible libraries".

**Comment by bignaux** [comment]
@edolstra : I first use it because i found it in another package and it is the nearest feature i was looking for, since i'm coming from gentoo. On that, i can make my package depend on specific use flag, and use flag globally. I use squashfuse on some kind of application bundle, and if i add an useless compression in my squashfuse libraries, it takes more space and dependencies to embed in the bundle. So i don't want "import all possible libraries". Btw, i've some more stuff to do on squashfuse, i imagine they release a new version after that.

**Comment by Profpatsch** [comment]
@Mic92 I felt like just merging it and ignoring my improvements was a bit … strange?

It was not really an optional improvement, since handling `null` everywhere is not exactly good or solid style for optional flags.

**Comment by bignaux** [comment]
@Profpatsch :  ceph pinentry xdelta shishi and wiredtiger use my solution so i thought it was a good solution since i find it better that your, that NO derivation are based on. Thanks to unmerge my patch, since i know there is always a maintener that have better idea on how do things. I get your point.

**Pull request merged**